### PR TITLE
Added const and oneOf functions in Producer module

### PR DESCRIPTION
**Pull request by eadred**
I've added two new functions to the Producer module, const and oneOf. The documentation I've added to the oneOf function explains the need/use for these.

The unit function can now also be redefined in terms of const.

I do have some property tests for these but I don't know where I should be adding test code.

### CODE REVIEW

File: src/Check/Producer.elm [hunk 1]
```diff
@@ -343,3 +351,35 @@ map : (a -> b) -> Producer a -> Producer b
...
+oneOf : List (Producer a) -> Producer a
+oneOf ps =
+    let -- Producer a -> Generator (a, Shrinker a)
+        -- Given a `Producer a` get a Generator that produces pairs consisting of
+        -- a value and a shrinker to apply to that value
+        toGenAndShrinker p = p.generator |> Random.map (\a -> (a, p.shrinker))
+        -- Generator (a, Shrinker a)
+        generator = ps |> List.map toGenAndShrinker |> Random.Extra.choices
+        -- Shrinker (a, Shrinker a)
+        shrinker (a, subshrinker) = subshrinker a |> Lazy.List.map (\a' -> (a', subshrinker))
```
Comment (eadred) [inline src/Check/Producer.elm]: When the Shrinker of the returned Producer receives a value we need to work out which Shrinker from the original list of producers we need to delegate to. This is why the constructed generator needs to return both the value and the shrinker of the producer that produced that value.

**Comment by mgold** [comment]
Thanks for the PR! I'll have time to review it fully in the next few days, but my biggest concerns on brief examination are, (1) this library is being folded into the next version of elm-test so I'm not sure what's worth merging, (2) are there real-world use cases for when you'd want to use these functions?

**Comment by eadred** [comment]
Thanks for getting back to me.

With regard to your second point the real world use case for oneOf is the ability to easily create Producers for discriminated union types by composing Producers for the underlying types used by the union. The example I have shown in the documentation is this one (and note that this also relies on the new const function too):

``` elm
type Example =
  Foo
  | Bar Int
  | Baz String

exampleProducer : Producer Example
exampleProducer =
  oneOf [ const Foo
        , int |> map Bar
        , string |> map Baz
        ]
```

This is in fact exactly the use case illustrated in the existing documentation for the Producer type, ie

``` elm
    type Question
        = Name String
        | Age Int


    question =
        let
            generator =
                Random.bool `Random.andThen` (\b ->
                    if b then
                        Random.map Name string.generator
                    else
                        Random.map Age (Random.int 0 120)
                 )

            shrinker question =
                case question of
                    Name n ->
                        Shrink.string n |> Shrink.map Name

                    Age i ->
                        Shrink.int i |> Shrink.map Age
        in
            Producer generator shrinker
```

Using the oneOf function is much more concise.

Incidentally the oneOf method was inspired by the F# FsCheck library _Gen.oneOf_ function - see [here](https://fscheck.github.io/FsCheck/TestData.html#Choosing-between-alternatives).

I'm not really sure what to suggest for your first point. I do believe these new functions have their uses (creating a producer for a discriminated union was one of the first things I found myself having to do when trying out elm-check) so it would be nice to see them integrated into the elm-check code, wherever that finds its home.

**Comment by eadred** [comment]
Ah! I've just noticed Producer.map doesn't map the Shrinker so I think I need to slightly redefine oneOf.

**Comment by eadred** [comment]
It looks like it is impossible to create a sensible shrinker for the result of oneOf. I've pushed a new change that just uses a noShrink explicitly instead.

This makes the function less comprehensive but still potentially useful. If you really did want a proper shrinker you'd have to implement this on an ad hoc basis as the current Producer documentation illustrates.

**Comment by mgold** [comment]
There's some discussion (project-fuzzball/test#14) about a new implementation of Fuzzer that would allow `map` to work sensibly (i.e. with shrinking), but also invalidate this implementation. I think it might be best to circle back to these ideas once that's implemented. I'd also want to see about matching the names with `random-extra` (and maybe add `frequency`). That's a concern for me because people are only willing to rewrite their tests so often, and I don't want to make work for people when we're going to be asking for a big rewrite in a few weeks.

**Comment by eadred** [comment]
Ok - no problem. I'll close this PR.

**Comment by mgold** [comment]
Thank you so much for contributing, sorry it didn't work out!

**Pull request closed without merging**